peerManager.go

```diff
@@ -0,0 +1,205 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache/ttl"
+	core "github.com/libp2p/go-libp2p-core"
+	"github.com/libp2p/go-libp2p-core/peer"
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var (
+	errNoDiscoverer = errors.New("discoverer isn't set")
+	errNoPeers      = errors.New("no peers are found")
+)
+
+var (
+	_timeout      = 10 * time.Minute
+	_connectRetry = 5
+)
+
+type peerManagerOpt func(cfg *peerManager)
+
+func withMaxPeers(num int) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.maxPeers = num
+	}
+}
+
+func withBlacklistTolerance(num int) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.blacklistTolerance = num
+	}
+}
+
```

Comment:
add a `func withBlackListTimeout(time.Duration) opt`

and add a timeout value into config